### PR TITLE
don't use `#[miri_run]` anymore, but execute the `main` function

### PR DESCRIPTION
**Pull request by oli-obk**
What do you think about this?

It somewhat hinders the ability to print the return value's memory of a function that returns a value.

I do have a minimal console stepper at hand that can be used in case someone wants to analyze memory (advantage: analyze any memory anywhere during miri's execution), I can add that to this PR if you want.

### CODE REVIEW

File: tests/compiletest.rs [hunk 1]
```diff
@@ -15,7 +15,7 @@ fn run_mode(mode: &'static str) {
             .expect("need to specify RUST_SYSROOT env var or use rustup or multirust")
             .to_owned(),
     };
-    let sysroot_flag = format!("--sysroot {}", sysroot);
+    let sysroot_flag = format!("--sysroot {} -Dwarnings", sysroot);
```
Comment (solson) [inline tests/compiletest.rs]: What is this for?

File: src/bin/miri.rs [hunk 2]
```diff
@@ -34,58 +37,46 @@ impl<'a> CompilerCalls<'a> for MiriCompilerCalls {
...
-                        }
-                        // FIXME: diverging functions can end up here in some future miri
-                        Err(e) => {
-                            report(tcx, &ecx, e);
-                            break;
-                        }
-                    }
+        if let map::Node::NodeItem(&Item { name, span, ref node, .. }) = tcx.map.get(id) {
+            if let ItemFn(..) = *node {
+                if name.as_str() == "main" {
```
Comment (solson) [inline src/bin/miri.rs]: This method doesn't account for `#[start]` which can make the start point something other than `main`.

File: tests/compiletest.rs (hunk 1 above)
Comment (oli-obk) [inline tests/compiletest.rs]: so our compile-fail tests don't contain unnecessary warnings anymore... like functions that are never called (those had `#[miri_run]`, but weren't called from `main`)

**Comment by solson** [comment]
I think this is good. I'd like to see the stepper, but don't necessarily add it to this PR.

File: tests/compiletest.rs (hunk 1 above)
Comment (solson) [inline tests/compiletest.rs]: Ah, okay. I guess rename `sysroot_flag` to `flags`.

File: src/bin/miri.rs (hunk 2 above)
Comment (eddyb) [inline src/bin/miri.rs]: `#[main]` is actually the thing here. See [this function](https://github.com/rust-lang/rust/blob/bf84f4e171fa0169af59881bc5d87ecc84d61523/src/librustc_trans/base.rs#L2356-L2361) for how to easily do this check.

**Comment by oli-obk** [comment]
works great for `#[start]`, `#[main]` and regular

**Comment by oli-obk** [comment]
all comments addressed

File: src/bin/miri.rs [hunk 3]
```diff
@@ -40,9 +36,12 @@ impl<'a> CompilerCalls<'a> for MiriCompilerCalls {
 
             let tcx = state.tcx.unwrap();
             let mir_map = state.mir_map.unwrap();
-            let (span, mir, def_id) = get_main(tcx, mir_map);
+
+            let (node_id, span) = state.session.entry_fn.borrow().expect("no main or start function found");
```
Comment (solson) [inline src/bin/miri.rs]: Huh, so this returns *either* the `#[main]` or `#[start]` function? That's a bit weird - we need to call them differently. Currently we're calling a `#[start]` without its two arguments, which I believe means they are left as undefined bytes.

File: tests/run-pass/start_fn.rs [hunk 4]
```diff
@@ -0,0 +1,6 @@
+#![feature(start)]
+
+#[start]
+fn foo(_nargs: isize, _args: *const *const u8) -> isize {
+    return 0;
```
Comment (solson) [inline tests/run-pass/start_fn.rs]: Nit: `return 0;` => `0`

File: src/bin/miri.rs [hunk 5]
```diff
@@ -40,9 +36,12 @@ impl<'a> CompilerCalls<'a> for MiriCompilerCalls {
...
             println!("found `main` function at: {:?}", span);
```
Comment (solson) [inline src/bin/miri.rs]: This should use a logging macro.

File: src/bin/miri.rs (hunk 3 above)
Comment (oli-obk) [inline src/bin/miri.rs]: fixed

File: src/bin/miri.rs (hunk 5 above)
Comment (oli-obk) [inline src/bin/miri.rs]: done

Comment (solson) [inline src/bin/miri.rs]: Still a `println` in the overall diff https://github.com/solson/miri/pull/23/files#diff-c3d602c5c8035a16699ce9c015bfeceaR41

Comment (oli-obk) [inline src/bin/miri.rs]: whoops, sorry, wanted to make that its own commit

**Comment by solson** [comment]
Thanks, I'll merge after CI.

**Comment by oli-obk** [comment]
CI is broken

**Comment by solson** [comment]
Yeah, looks like we have some sysroot problems there.
